### PR TITLE
Ryan/dev

### PR DESCRIPTION
**Pull request by Ryan-Rhys**
Opening for feedback. 

Added `bit_vector_kernel.py` module.

- Abstract base class for bit vector kernels
- implementation of Tanimoto kernel
- Tests for Tanimoto distance metric
- TODOs during review: further tests, design discussion

### CODE REVIEW

File: tests/test_bit_kernels.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    """
+    x1 = torch.ones((2, 2))
+    x2 = 2*torch.ones((2, 2))
+    dist_object = BitDistance()
+    tan_similarity = dist_object._tan_similarity(x1, x2, postprocess=False)
+
+    assert torch.allclose(tan_similarity, torch.tensor(0.6666666666666))
+
+
+# GPflow Tanimoto kernel implementation below
```
Comment (gck25) [inline tests/test_bit_kernels.py:41]: This looks different to the Tanimoto implementation in bit_vector_kernel.py.  What is the rationale behind including this here?

**Review by gck25 — COMMENTED** [review]
Really well-coded.  Just one question concerning the test file.

File: gprotorch/kernels/bit_vector_kernel.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            :attr: `x1_eq_x2` (bool):
+                Is x1 equal to x2
+            :attr: `metric` (str):
+                String specifying the similarity metric. One of ['tanimoto']
+
+        Returns:
+            (:class:`Tensor`, :class:`Tensor) corresponding to the similarity matrix between `x1` and `x2`
+        """
+
+        if metric not in ['tanimoto']:
```
Comment (JuliusSchwartz) [inline gprotorch/kernels/bit_vector_kernel.py:40]: Why not `if metric != 'tanimoto'` ?

File: gprotorch/kernels/bit_vector_kernel.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            raise RuntimeError('Similarity metric not supported. Available options are \'tanimoto\'')
+
+        if metric == 'tanimoto':
```
Comment (JuliusSchwartz) [inline gprotorch/kernels/bit_vector_kernel.py:43]: Do we need this check? Looks like we're already raising an error if this isn't the case

File: gprotorch/kernels/bit_vector_kernel.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        r"""
+        This is a helper method for computing the bit vector similarity between
+        all pairs of points in x1 and x2.
+
+        Args:
+            :attr:`x1` (Tensor `n x d` or `b1 x ... x bk x n x d`):
+                First set of data.
+            :attr:`x2` (Tensor `m x d` or `b1 x ... x bk x m x d`):
+                Second set of data.
+            :attr:`diag` (bool):
```
Comment (JuliusSchwartz) [inline gprotorch/kernels/bit_vector_kernel.py:114]: Is `diag` supposed to be one of the args to this function?

File: gprotorch/kernels/bit_vector_kernel.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+    This base :class:`BitKernel` class does not include a lengthscale parameter
+    :math:`\Theta`, in contrast to many common kernel functions.
+
+    Base Attributes:
+    :attr:`metric` (str):
+        The similarity metric to use. One of ['tanimoto'].
+    """
+    def __init__(self, metric='', **kwargs):
+        super(BitKernel, self).__init__(**kwargs)
```
Comment (JuliusSchwartz) [inline gprotorch/kernels/bit_vector_kernel.py:90]: I've been out of the python world for a bit - is it not possible and easier to do `super().__init__(<stuff>)`? This is what you've done on line 17

File: tests/test_bit_kernels.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+
+@pytest.mark.parametrize("x1, x2", [
+    (torch.ones((2, 2)), torch.ones((2, 2))),
+    (2*torch.ones((2, 2)), 2*torch.ones((2, 2))),
+    (10*torch.ones((2, 2)), 10*torch.ones((2, 2))),
+    (10000*torch.ones((2, 2)), 10000*torch.ones((2, 2))),
+    (0.1*torch.ones((2, 2)), 0.1*torch.ones((2, 2)))
+])
+def test_tanimoto_similarity_with_equal_inputs(x1, x2):
```
Comment (JuliusSchwartz) [inline tests/test_bit_kernels.py:23]: Does this test need two args if you're only calling it when both args are the same (and the test would fail if they were different)?

File: gprotorch/kernels/bit_vector_kernel.py (hunk 2 above)
Comment (leojklarner) [reply gprotorch/kernels/bit_vector_kernel.py:40]: @JuliusSchwartz We were thinking about adding some other bit vector kernels, see e.g. table 6 in [this message](https://gprotorch.slack.com/archives/C01MLAZBRPH/p1616598648041100) in the Slack.

File: gprotorch/kernels/bit_vector_kernel.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        """
+
+        if metric not in ['tanimoto']:
+            raise RuntimeError('Similarity metric not supported. Available options are \'tanimoto\'')
+
+        if metric == 'tanimoto':
+
+            # Tanimoto distance is proportional to (<x, y>) / (||x||^2 + ||y||^2 - <x, y>) where x and y are bit vectors
+
+            x1_norm = x1.pow(2).sum(dim=-1, keepdim=True)  # Compute the squared L2-norm of each datapoint
```
[the remark this answers is not in the document]
Comment (leojklarner) [reply gprotorch/kernels/bit_vector_kernel.py:47]: We'd need it if we used count vectors, but I'm not sure the Tanimoto similarity is a valid metric for these, as you can construct an example where a molecule is more similar to another molecule than to itself.

**Review by leojklarner — COMMENTED** [review]
Great job! I can't remember what I wanted to refactor last time, but I couldn't think of anything this time so it couldn't have been important.

File: gprotorch/kernels/bit_vector_kernel.py (hunk 2 above)
Comment (Ryan-Rhys) [reply gprotorch/kernels/bit_vector_kernel.py]: I'm planning on extending this list with all the different metrics we're going to use e.g. MinMax, Tanimoto etc. I'll add a comment to that effect so it's clear!

File: gprotorch/kernels/bit_vector_kernel.py (hunk 3 above)
Comment (Ryan-Rhys) [reply gprotorch/kernels/bit_vector_kernel.py]: So this branch will take us into the Tanimoto metric, I'll add further branches for the other metrics we introduce

File: gprotorch/kernels/bit_vector_kernel.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ryan-Rhys) [reply gprotorch/kernels/bit_vector_kernel.py]: Nice point! I used Tanimoto on count vectors for the photoswitch dataset and it worked OK. Might be worth leaving in in the case that users want to use Tanimoto on count vectors?

File: gprotorch/kernels/bit_vector_kernel.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+            denominator = x1_norm + x2_norm.squeeze(-1) - cross_product
+
+            res = (cross_product/denominator)
+
+            # if x1 and x2 are equal than the Tanimoto similarity is 1
+            if x1_eq_x2 and not x1.requires_grad and not x2.requires_grad:
+                res.diagonal(dim1=-2, dim2=-1).fill_(1)
+
+            # Zero out negative values
+            res.clamp_min_(0)
```
[the remark this answers is not in the document]
Comment (Ryan-Rhys) [reply gprotorch/kernels/bit_vector_kernel.py]: Yep, it's for numerical reasons! They make use of it in the main GPyTorch repo

File: gprotorch/kernels/bit_vector_kernel.py (hunk 5 above)
Comment (Ryan-Rhys) [reply gprotorch/kernels/bit_vector_kernel.py]: Made this change!

File: gprotorch/kernels/bit_vector_kernel.py (hunk 4 above)
Comment (Ryan-Rhys) [reply gprotorch/kernels/bit_vector_kernel.py]: I've taken it out, it's included in the base Kernel class of GPyTorch but I'm not sure what kind of applications it's needed for

File: tests/test_bit_kernels.py (hunk 6 above)
Comment (Ryan-Rhys) [reply tests/test_bit_kernels.py]: True, I've just left it as x1 and x2 as hopefully it'll be clearer what's going on to anyone who might look at the tests!
